craft_parts/executor/collisions.py

```diff
@@ -28,22 +28,60 @@
 from craft_parts.permissions import Permissions, permissions_are_compatible
 
 
-def check_for_stage_collisions(part_list: List[Part]) -> None:
+def check_for_stage_collisions(
+    part_list: List[Part], partitions: Optional[List[str]]
+) -> None:
     """Verify whether parts have conflicting files to stage.
 
-    :param part_list: The list of parts to be tested.
+    If the partitions feature is enabled, then check if parts have conflicting files to
+        stage for each partition.
+    If the partitions feature is disabled, only check for conflicts in the default
+        stage directory.
+
+    :param part_list: The list of parts to check.
+    :param partitions: An optional list of partition names.
+
+    :raises PartConflictError: If conflicts are found.
+    :raises FeatureError: If partitions are specified but the feature is not enabled or
+        if partitions are not specified and the feature is enabled.
+    """
+    if partitions and not Features().enable_partitions:
+        raise errors.FeatureError(
+            "Partitions specified but partitions feature is not enabled."
+        )
+
+    if partitions is None and Features().enable_partitions:
+        raise errors.FeatureError(
+            "Partitions feature is enabled but no partitions specified."
+        )
+
+    if partitions:
+        for partition in partitions:
+            _check_for_stage_collisions_per_stage_directory(part_list, partition)
+    else:
+        _check_for_stage_collisions_per_stage_directory(part_list, None)
+
+
+def _check_for_stage_collisions_per_stage_directory(
+    part_list: List[Part], partition: Optional[str]
```

Comment:
Passing the path directly as an argument to the function could make the internal logic easier to follow, and avoid passing `None` to be used as a dictionary key.

Reply:
I can not figure out how to implement your suggestion!  

I can't pass a single directory to this function; it needs more than just a single path.  It needs `part_install_dir[partition]` for all parts and it also needs the partition name for calls like `filesets.migratable_filesets(stage_fileset, srcdir, partition)`.

Reply:
You're right, you're iterating over all parts there. Thanks for the clarification.